fixed: Try to recache image in case cached texture loading fails

arnova opened the pull request:
Up till now we didn't handle failed texture loading for cached images. This could occur in cases where the locally cached file is missing (or corrupt) and possibly other scenarios.

Commit:
changed: Refactor ImageLoader to improve error handling when e.g. texture loading fails

arnova commented:
jenkins build this please

arnova commented:
I've revived this old PR by refactoring the LargeTextureManager as suggested by @jmarshallnz . Please comment.

mkortstiege commented on xbmc/GUILargeTextureManager.cpp:
```diff
@@ -57,27 +57,33 @@ bool CImageLoader::DoWork()
-  return true;
+
+  // not in our texture cache or it failed to load from it, so try and load directly and then cache the result
+  loadPath = CTextureCache::Get().CacheImage(texturePath, &m_texture);
```
loadPath can be dropped here

arnova commented on xbmc/GUILargeTextureManager.cpp:
```diff
@@ -57,27 +57,33 @@ bool CImageLoader::DoWork()
-      return true; // we're done
-  }
   if (!m_use_cache || !loadPath.empty())
   {
```
@mkortstiege : Btw. you see the logic in the above if-statement above? Seems there's no point doing this block if loadPath is empty (so with it being empty and m_use_cache = 0 we would enter this block)?

mkortstiege replied:
Right, looks indeed fishy. I'd remove loadpath from that check as `LoadFromFile` will return false anyways in case it's empty and we're told to not read from cache.

arnova commented:
@mkortstiege : I've updated the PR. I also changed the if-logic we previously talked about. I think I got it right now (after some (long) thinking). I'm quite pleased with it now and I'm pretty confident I got all cases covered now.

Please review and let me know. Perhaps @jmarshallnz is still reading mail and if so, likes the comment as well?

arnova commented:
jenkins build this please

Commit:
fixed: Declaration order + wrong prototype (cosmetics)

arnova commented:
jenkins build this please

mkortstiege commented:
Looks good.

arnova commented:
If not objections, will merge tonight.

ace20022 commented on xbmc/TextureCache.cpp:
```diff
@@ -176,6 +167,15 @@ std::string CTextureCache::CacheImage(const std::string &image, CBaseTexture **t
+
+  return !path.empty();
+}
+
```
Why have you move this method down?

ace20022 replied:
Just read the commit message, nvm ;)
